### PR TITLE
Pallet-restricted-tokens [New]

### PR DESCRIPTION
**Pull request by mustermeiszer**
This basically copies the approach made in the PR here https://github.com/open-web3-stack/open-runtime-module-library. But as we want to have the idea in for the audit, this PR provides a pallet that exposes the same Call-interface as ORML-tokens but has the trait `PreConditions` to check for any conditions on a transfer. 

The runtimes are then free to choose, which checks they want to run. We do check for permissions of both sender and receiver in case someone tries to transfer a restricted token. 

Edit: 
The PR is extended for the following logic: Pallet-restricted-tokens acts as a proxy for all token related actions including traits. For this reason the pallet implements
- fungibles::{Mutate, Transfer, Inspect, MutateHold, InspectHold}
 - fungible::{Mutate, Transfer, Inspect, MutateHold, InspectHold}
 - Currency, LockableCurrency, ReservableCurrency

### CODE REVIEW

**Review by vedhavyas — APPROVED** [review]
LGTM!
Will benchmarks be part of another PR. I suppose that will be mostly copy paste from orml mybe ?

**Comment by vedhavyas** [comment]
@mustermeiszer can we also adjust loan and pool pallet usage of ORML to restricted tokens and fungible traits or will this change be in a follow-up pr ?

File: pallets/restricted-tokens/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen;
+
+		type CurrencyId: Parameter + Member + Copy + MaybeSerializeDeserialize + Ord + TypeInfo;
+
+		type Fungibles: Mutate<Self::AccountId>
+			+ Transfer<Self::AccountId>
+			+ Inspect<Self::AccountId, AssetId = Self::CurrencyId, Balance = Self::Balance>
+			+ Unbalanced<Self::AccountId>
+			+ MutateHold<Self::AccountId>;
```
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/restricted-tokens/src/lib.rs:87]: Are those fungible ones? Will add those bounds.

**Comment by mustermeiszer** [comment]
Yes, benchmarking will be part of another pr. I just wanted this in for the audit. And I will adjust pool and loan with the right fungible traits and wire this in.

**Review by offerijns — APPROVED** [review]

**Review by vedhavyas — COMMENTED** [review]
I'm still on the edge when it comes to Native fungibles being part of this pallet.
- This opens up to us being maintaining 
- Exchanges and Clients needs to rework on the their integrations which is a pain

I would prefer if we just the fungibles and not the native for now. We should Ideally revisit this down the line. Thoughts @lucasvo @offerijns

**Comment by mustermeiszer** [comment]
> I'm still on the edge when it comes to Native fungibles being part of this pallet.

This opens up to us being maintaining
Exchanges and Clients needs to rework on the their integrations which is a pain
I would prefer if we just the fungibles and not the native for now. We should Ideally revisit this down the line. Thoughts @lucasvo @offerijns

As discussed offline I will merge this as is.

**Pull request merged**